candidclient/identity.go

```diff
@@ -72,60 +73,55 @@ func (id *usernameIdentity) Domain() string {
 }
 
 type useridIdentity struct {
-	client   *Client
-	userID   string
-	username string
+	client *Client
+	user   params.User
 }
 
 // Username implements Identity.Username.
 func (id *useridIdentity) Username() (string, error) {
-	if id.username != "" {
-		return id.username, nil
+	if id.user.Username != "" {
+		return string(id.user.Username), nil
 	}
 
 	ctx := context.Background()
-	usernames, err := id.client.QueryUsers(ctx, &params.QueryUsersRequest{
-		ExternalID: id.userID,
+	user, err := id.client.GetUserWithID(ctx, &params.GetUserWithIDRequest{
+		UserID: id.user.ExternalID,
 	})
 	if err != nil {
 		return "", errgo.Mask(err)
 	}
-	if len(usernames) == 1 {
-		id.username = usernames[0]
-	}
-	return id.username, nil
+	id.user = *user
+	return string(id.user.Username), nil
 }
 
 // Groups implements Identity.Groups.
 func (id *useridIdentity) Groups() ([]string, error) {
-	username, err := id.Username()
+	_, err := id.Username()
```

Comment:
Would it be more explicit if both `Groups()` and `Username()` called an internal function that would populate `id.user` ?

Reply:
maybe, I don't think it would make it much simpler though.